third_party/amd/lib/TritonAMDGPUToLLVM/SchedInstructions.cpp

```diff
@@ -181,14 +199,108 @@ struct InsertInstructionSchedHints
     MLIRContext *ctx = &getContext();
     ModuleOp mod = getOperation();
 
-    mod->walk([ctx](triton::DotOp dot) {
-      if (dyn_cast<mlir::scf::ForOp>(dot->getParentOp())) {
+    mod->walk([this, ctx](triton::DotOp dot) {
+      if (auto forOp = dyn_cast<scf::ForOp>(dot->getParentOp())) {
         mlir::OpBuilder rewriter(ctx);
         rewriter.setInsertionPointAfter(dot);
         rewriter.create<triton::amdgpu::InstructionSchedHint>(dot->getLoc());
+        anotateDotOperands(forOp);
       }
     });
   }
+
+  template <typename OpType>
+  llvm::SmallVector<OpType> getUsersOfType(Value value) const {
+    llvm::SmallVector<OpType> concreteUsers;
+    for (auto user : value.getUsers()) {
+      if (auto concreteUser = llvm::dyn_cast<OpType>(user))
+        concreteUsers.push_back(concreteUser);
+    }
+    return concreteUsers;
+  }
+
+  Value rewindConvertLayoutOps(Value value) const {
+    auto convertOps = getUsersOfType<triton::gpu::ConvertLayoutOp>(value);
+    while (!convertOps.empty()) {
+      assert(convertOps.size() == 1);
+      value = convertOps.begin()->getResult();
+      convertOps = getUsersOfType<triton::gpu::ConvertLayoutOp>(value);
+    }
+    return value;
+  }
+
+  void anotateDotOperands(scf::ForOp forOp) const {
+    llvm::SmallVector<triton::LoadOp> loadOps;
+    forOp.walk(
+        [&loadOps](triton::LoadOp loadOp) { loadOps.push_back(loadOp); });
+
+    ValueRange yieldedValues = forOp.getYieldedValues();
+    auto initArgs = forOp.getRegionIterArgs();
+
+    MLIRContext *ctx = forOp->getContext();
+    mlir::OpBuilder rewriter(ctx);
+
+    for (auto loadOp : loadOps) {
+      Value loadResult = loadOp.getResult();
+
+      // rewind till the first carried loop iteration regarding `tt.load`
```

Comment:
Style: captialize the first word and end comment with `.`

Reply:
Done